ouroboros-consensus/docs/VersioningSchemeDecision.md

```diff
@@ -0,0 +1,422 @@
+The most likely way we can imagine a dev version might end up in a repository would be if a developer/user who doesn't know any better builds from source and installs a dev version into their machine's local package database.
+That single risk seems manageable.
+
+This suggests the following proposal.
+- Each release version is `A.B.C`.
+- Each dev version is `A.B.C.D`.
+- Any PR that alters a package in a way that should affect the next release version should update its version as follows.
+    - If the version was `A.B.C` before the PR, bump it to `A.B.C.0`.
+    - If the version was `A.B.C.D` before the PR, bump it to `A.B.C.(D+1)`.
+        - This rule isn't fundamental, but it ensures that every PR touches the version, which in turn ensures `git` will raise merge conflicts for backports/rebases/cherry-picking/etc.
+          Those conflicts force the developer to make adjustments to versions.
+          Those conflicts are necessary and also trivial to resolve but would be easy to overlook without this rule leveraging `git` to remind us.
+          Additionally, this means that no two code-changing commits in the subset of a single branch will declare the same version---or set of versions if the repo has multiple packages---which seems intuitive.
+        - Similar conflicts will also require updating open version-altering PRs after each version-altering PR is merged.
+          Con: that will likely be tedious.
+          Pro: it will likely lead to more linear integration branch history: you might was well rebase while you're updating the version bumps.
+        - If you want `D` to exactly count the number of relevant PRs since the previous release, then you'll need to require every merge is a [fast-forward merge](https://git-scm.com/docs/git-merge#_fast_forward_merge); see [`git merge --ff-only`](https://git-scm.com/docs/git-merge#Documentation/git-merge.txt---ff-only).
+- Cut a new release as follows.
+    - Merge a fresh PR that doesn't alter the package but does bump dev version `A.B.C.D` to the next release version, depending on the contents of the relevant PRs since the previous release:
+      `(A+1).0.0` if there were breaking changes, else `A.(B+1).0` if there were some new features, else `A.B.(C+1)`.
+      (Note that `.D` is present, so at least one version-influencing change was made.
+      If the version in the commit was still `A.B.C`, why are you announcing a new release?)
+    - Announce that commit as a release of the packages that have changed since the previous release.
+      (In a monorepo, there's just the one.
+      But in a polyrepo, there may be packages that haven't changed since the previous release.)
+
+## Proposal EasierRedimensional
+
+There is a variant of the above that means only release PRs alter the version, which is more scalable.
```

Comment:
I would say this this is instead a variant of RisingEdge that avoids one of the cons (that main and release versions look the same). It's RisingEdge, except that just after you do the release you bump to a "dev version" by adding zeros until you get to the fourth zero.

Reply:
RisingEdge continues to be my preference for simplicity, this seems like an acceptably small tweak.

Reply:
Just to clarify:

> except that just after you do the release you bump to a "dev version" by adding zeros until you get to the fourth zero.

Every release version has three components (or 4, for PVP). So you only ever add one `.0`.

Reply:
Ah right. It would also work if you didn't do that, and you could release e.g. just `A.B`. Then I guess you would say that a "dev version" is one with a trailing 0. But your way is also fine.